File: schedule.py
#from bottle import route, run, request, install, static_file, template, redirect
import json
from bottle import *
#import bottle
from bottle_sqlite import SQLitePlugin

def get_schedules(db):
  row = db.execute('SELECT * from schedule').fetchall()
  if row:
    #return template('home', schedules=row)
    results = []
    for data in row:
        results.append(dict(data))
    return json.dumps(results)
  return null

@route('/schedule', method='GET')
def schedules(db):
    response.content_type = 'application/json'
    return get_schedules(db)

@route('/schedule', method='POST')
def create_schedule(db):
    if not request.json or not 'name' in request.json:
        abort(400)
    row = db.execute('SELECT * from schedule').fetchall()
    if row:
        id = row[-1]['id'] + 1
    else:
        id = 1
    import datetime
    deadline = str(datetime.datetime.now()+datetime.timedelta(days=request.json.get('deadline', 0)))
    schedule = {
        'id': id,
        'name': request.json['name'],
        'deadline': deadline,
        'description': request.json.get('description', ""),
        'complete': 0
    }
    db.execute("INSERT INTO schedule VALUES (?, ?, ?, ?, ?)", (schedule['id'], schedule['name'], schedule['deadline'], schedule['description'], schedule['complete']))
    db.commit()
    return HTTPResponse(status=201, body=json.dumps(schedule))

@route('/schedule/<idnum>', method='GET')
def schedule(idnum, db):
  row = db.execute('SELECT * from schedule where id=?', [idnum]).fetchone()
  if row:
    response.content_type = 'application/json'
    return json.dumps(dict(row))
  return HTTPError(404, "Page not found")

@route('/schedule/<idnum>', method='DELETE')
def delete_schedule(idnum, db):
  delete = db.execute('DELETE from schedule where id=?', [idnum])
  db.commit()
  return ''

def put_data(db, numid, column, data):
  # Put data
  db.text_factory = str
  row = db.execute('SELECT * from schedule where id=?', [numid]).fetchone()
  if row:
    updatestr = 'UPDATE schedule set %s=? where id=?' % column
    print column, data
    update_data = db.execute(updatestr, [data, numid])
    db.commit()
    row = db.execute('SELECT * from schedule where id=?', [numid]).fetchone()
    if row:
      return json.dumps(dict(row))
  return HTTPError(404, "Data not found")

@route('/schedule/<idnum>/name', method='PUT')
def put_name(idnum, db):
  if not request.json or not 'name' in request.json:
      abort(400)
  name = request.json['name']
  response.content_type = 'application/json'
  return put_data(db, idnum, column='name', data=name)

@route('/schedule/<idnum>/description', method='PUT')
def put_description(idnum, db):
  if not request.json or not 'description' in request.json:
      abort(400)
  description = request.json['description']
  response.content_type = 'application/json'
  return put_data(db, idnum, column='description', data=description)

@route('/schedule/<idnum>/deadline', method='PUT')
def put_deadline(idnum, db):
  if not request.json or not 'deadline' in request.json:
      abort(400)
  # deadline format is %d/%m/%y %H:%M
  import datetime
  deadline = datetime.datetime.strptime(request.json.get('deadline', datetime.datetime.now()), "%d/%m/%y %H:%M")
  response.content_type = 'application/json'
  return put_data(db, idnum, column='deadline', data=deadline)

@route('/schedule/<idnum>/complete', method='PUT')
def put_complete(idnum, db):
  row = db.execute('SELECT * from schedule where id=?', [idnum]).fetchone()
  if row:
    complete = row['complete']
    # toggle complete flag
    if complete == 0:
      complete = 1
    else:
      complete = 0
    response.content_type = 'application/json'
    return put_data(db, idnum, column='complete', data=complete)
  return HTTPError(404, "Data not found")

if __name__ == "__main__":
  sqlite = SQLitePlugin(dbfile='schedule.db')
  install(sqlite)
 
  run(host='<HOST IP>', port=<PORT>)

File: schedule.js
var express = require('express');
var bodyParser = require('body-parser');
var app = express();
var sqlite3 = require('sqlite3').verbose();
var fs = require('fs');
var csv = require('csv');
var router = express.Router();

app.use(bodyParser.json());

fs.stat('schedule.db', function(err, stats) {
  if (err) {
    var db = new sqlite3.Database('schedule.db');
    console.log('schedule.db not found');
    db.serialize(function() {
      db.run("CREATE TABLE schedule ( id INTEGER PRIMARY KEY, name TEXT, deadline TIMESTAMP, description TEXT, complete INTEGER )");
      var csvfile = fs.createReadStream('sample_schedule.csv');
      var index = 0;
      csvfile.pipe(csv.parse())
        .pipe(csv.transform(function(record) {
          index = index + 1;
          var datetime = new Date();
          console.log(record);
          datetime.setDate (datetime.getDate() + index);
          db.run("INSERT INTO schedule VALUES (?, ?, ?, ?, ?)", index, record[0], datetime.toISOString(), record[1], 0);
        }));
    });
  }
});

var db = new sqlite3.Database('schedule.db');

function get_schedules(res) {
  db.all('SELECT * from schedule',function(err,rows) {
    var results = new Array();
    rows.forEach(function (row) {
      results.push(row);
    });
    res.json(results);
  });
};

router.get('/', function(req, res) {
  get_schedules(res);
});

router.post('/', function(req,res) {
  if (!req.body || !req.body.name) {
    res.sendStatus(400);
  }
  else {
    db.all('SELECT * from schedule', function(err, rows) {
      console.log(rows);
      var id = rows[rows.length-1].id+1;
      var datetime = new Date();
      datetime.setDate (datetime.getDate() + req.body.deadline);
      var schedule = {
        'id': id,
        'name': req.body.name,
        'deadline': datetime.toISOString(),
        'description': req.body.description,
        'complete': 0
      };
      db.run("INSERT INTO schedule VALUES (?, ?, ?, ?, ?)", schedule.id, schedule.name, schedule.deadline, schedule.description, schedule.complete);
      res.json(schedule);
    });
  }
});

router.get('/:id', function(req, res) {
  db.get('SELECT * from schedule where id=?', req.params.id, function(err, row) {
    if (err || !row) {
      res.sendStatus(404);
    }
    else {
      res.json(row);
    }
  });
});

router.delete('/:id', function(req, res) {
  db.run('DELETE from schedule where id=?', req.params.id);
  res.send(true);
});

function put_data(db, numid, column, data) {
  // Put data
  var data;
  db.serialize(function() {
    db.get('SELECT * from schedule where id=?', numid, function(err, testrow) {
      if (err || !testrow) {
        data = null;
      }
      else {
        console.log("column="+column+", data="+data+", numid="+numid);
        db.run('UPDATE schedule set '+column+'=? where id=?', data, numid);
        db.get('SELECT * from schedule where id=?', numid, function(err, row) {
          if (err || !row) {
            data = null;
          }
          else {
            data = row;
          }
        });
      }
    });
  });
  return data;
};

router.put('/:id/name', function(req, res) {
  if (!req.body || !req.body.name) {
    res.sendStatus(400);
  }
  else {
    var data = put_data(db, req.params.id, 'name', req.body.name);
    if (data == null)
      res.sendStatus(404);
    else
      res.json(data);
  }
});

router.put('/:id/description', function(req, res) {
  if (!req.body || !req.body.description) {
    res.sendStatus(400);
  }
  else {
    var data = put_data(db, req.params.id, 'description', req.body.description);
    if (data == null)
      res.sendStatus(404);
    else
      res.json(data);
  }
});

router.put('/:id/deadline', function(req, res) {
  if (!req.body || !req.body.deadline) {
    res.sendStatus(400);
  }
  else {
    // deadline format is %d/%m/%y %H:%M
    var datetime = new Date();
    datetime.setDate (datetime.getDate() + req.body.deadline);
    var data = put_data(db, req.params.id, 'deadline', datetime.toISOString());
    if (data == null)
      res.sendStatus(404);
    else
      res.json(data);
  }
});

router.put('/:id/complete', function(req, res) {
  db.get('SELECT * from schedule where id=?', req.params.id, function(err, row) {
    if (err || !row) {
      res.sendStatus(404);
    }
    else {
      // toggle complete flag
      var complete = (row.complete==0)?1:0;
      var data = put_data(db, req.params.id, 'complete', complete);
      if (data == null)
        res.sendStatus(404);
      else
        res.json(data);
    }
  });
});

app.use('/schedule', router);

app.listen(5000, function() {
  console.log("Live at Port 5000");
});
